### PR TITLE
Update libcxxabi to llvm 9.0.0

### PR DESCRIPTION
**Pull request by sbc100**
This change was generate by checking out llvm 9.0.0 and running
`./system/lib/update_libcxxabi.py ../llvm-project/` then searching
the diff for EMSCRIPTEN and looking at the readme file to ensure our
local changes were preserved.

### CODE REVIEW

File: system/lib/libc/musl/arch/emscripten/bits/syscall.h [hunk 1]
```diff
@@ -573,7 +573,7 @@
 #define SYS_getdents64		220
 #define SYS_fcntl64		221
 /* 223 is unused */
-#define SYS_gettid		224
+/*#define SYS_gettid		224*/
```
Comment (kripken) [inline system/lib/libc/musl/arch/emscripten/bits/syscall.h]: these changes don't look related to libcxxabi? are they are workaround for something?

File: system/lib/libcxxabi/include/cxxabi.h [hunk 2]
```diff
@@ -86,7 +85,7 @@ extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_NORETURN void __cxa_pure_virtual(void);
 extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_NORETURN void __cxa_deleted_virtual(void);
 
 // 3.3.2 One-time Construction API
-#if defined(__arm__) || defined (__EMSCRIPTEN__)
+#if defined(_LIBCXXABI_GUARD_ABI_ARM)
```
Comment (kripken) [inline system/lib/libcxxabi/include/cxxabi.h]: looks like we lost an `__EMSCRIPTEN__` guard here?

File: system/lib/libc/musl/arch/emscripten/bits/syscall.h (hunk 1 above)
Comment (sbc100) [reply system/lib/libc/musl/arch/emscripten/bits/syscall.h]: Oh yeah sorry, there is code in libcxxabi the does `if defined(SYS_gettid)`.     I guess maybe we should look at whether we need any of these defines.  Certainly these two are not supposed syscalls.

https://github.com/llvm-mirror/libcxxabi/blob/master/src/cxa_guard_impl.h#L121

Do we even need any of these macros?

File: system/lib/libcxxabi/include/cxxabi.h (hunk 2 above)
Comment (sbc100) [reply system/lib/libcxxabi/include/cxxabi.h]: See the comment in tools/system_libs.py.  I defined _LIBCXXABI_GUARD_ABI_ARM on the command line.   We can remove that once we roll to 10.0.0 since its defined upsteam now for wasm:
https://github.com/llvm-mirror/libcxxabi/blob/master/include/__cxxabi_config.h#L75

File: system/lib/libc/musl/arch/emscripten/bits/syscall.h (hunk 1 above)
Comment (kripken) [reply system/lib/libc/musl/arch/emscripten/bits/syscall.h]: I think they are used for musl syscalls? (the ones we actually use)

Sounds good to remove ones we don't need. How about doing it like the existing `/* X is unsed */`?

File: system/lib/libcxxabi/include/cxxabi.h (hunk 2 above)
Comment (kripken) [reply system/lib/libcxxabi/include/cxxabi.h]: Seems like a reasonable workaround, I guess, but confusing since we don't use the ARM ABI (or do we?). But I guess there's no simpler way?

File: tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.funcs [hunk 3]
```diff
@@ -426,6 +433,7 @@ $___uflow
 $___uselocale
 $___vfprintf_internal
 $___wasi_syscall_ret
+$_abort_message
```
Comment (kripken) [inline tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.funcs]: this, and the other additions here, look bad to me - I think we should investigate if there isn't something we can ifdef away in this new code. OTOH I see the code size didn't need to change, so it's within a small range, but still.

File: system/lib/libcxxabi/include/cxxabi.h (hunk 2 above)
Comment (sbc100) [reply system/lib/libcxxabi/include/cxxabi.h]: I'm not changing this behaviour as part of this change.  We were doing this both before and after.   IIRC we did do it (and have always done it) this way to save space, and I landed this change upstream llvm to be consistent with out ABI.   

See: https://reviews.llvm.org/D64961

File: tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.funcs (hunk 3 above)
Comment (sbc100) [reply tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.funcs]: I tried to dig into these but didn't get very far.   It just looks like a normal amount of churn to me.   If its not a large code size increase I'm tempted to let it go.   But I will try to dig in a little more.

Comment (kripken) [reply tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.funcs]: Might be good to at least measure the code size regression here in a precise way.

File: system/lib/libc/musl/arch/emscripten/bits/syscall.h (hunk 1 above)
Comment (sbc100) [reply system/lib/libc/musl/arch/emscripten/bits/syscall.h]: I'm going to work on followup change to cleanup this file and remove any syscalls we don't implement.

**Comment by sbc100** [comment]
I looked into the symbol differences.  They all seems to be related the main change here which is the generic code for guarding C++ global value initialization.

For fastcomp the new C symbols are:

- __overflow - pulled in due to putc_unlocked which is macro used by ..
- _fputc  - pulled in due to printf
- _vfprintf  - pulled in due to abort_message
- _abort_message - pulled in due to cxa_guard_impl.h changes for such things as `ABORT_WITH_MESSAGE("__cxa_guard_acquire detected recursive initialization");`

The new C++ symbols are all related to GuardObject which is the new way C++ guard are working.

I noticed that the _abort_message stuff seems to be gates on `ifndef NDEBUG`.   We should presumably be building with NDEBUG?  Maybe I'll try to land that change first.

**Comment by sbc100** [comment]
Nope.. we are already defining NDEBUG..

**Comment by kripken** [comment]
How about replacing `ABORT_WITH_MESSAGE` with a call to the JS `abort(..)`, which would print the argument, then abort, with no cost in code size?

File: system/lib/libc/musl/arch/emscripten/bits/syscall.h (hunk 1 above)
Comment (sbc100) [reply system/lib/libc/musl/arch/emscripten/bits/syscall.h]: This change is no longer part of this CL.

**Comment by sbc100** [comment]
OK, I converted libc++abi's abort_mesasge to use emscripten's emscripten_log + abort functions.   It means more JS dependencies:

```
__emscripten_traverse_stack
__formatString
__reallyNegative
_emscripten_log
_emscripten_log_js
_emscripten_get_callstack_js
convertI32PairToI53
convertU32PairToI53
```

So I'm not sure its a win since it also hurts STANDALONE_WASM builds.   I don't feel strongly either wsy.

**Comment by sbc100** [comment]
Actually those symbols seems to be added the build only for fastcomp.  

For the llvm backend build the change in the size of the resulting wasm is less than 1k

**Comment by sbc100** [comment]
Its 1k out of 270k .. not sure if that changes the calculus?

**Comment by sbc100** [comment]
For the wasm binary built as part of `test_metadce_cxx_noexcept` that is.

**Comment by kripken** [comment]
No worries about landing. I can take a look at the size afterwards.

**Comment by kripken** [comment]
Looks like the abort message is a small part of the increase. It's mostly in the new libc++abi logic that checks for recursive initialization, adding a bunch more bookkeeping. Removing all that might be possible, but not pretty, so probably we just need to live with this.

**Comment by kripken** [comment]
Adding NDEBUG checks around those things can help by 100-200 bytes, so just a small part of the overall refactor. And I'm assuming they had a reason not to do that - maybe it's part of the spec to abort there perhaps.
